fpgalib/dac.py

```diff
@@ -1477,19 +1477,56 @@ def make_jump_table_entry(cls, name, arg):
     def make_jump_table(cls, jt_entries, counters=None, start_address_ns=0):
         """Make a jump table out of the given entries and counters.
 
+        We verify that jump table commands are not executed too frequently. We
+        check when our current command executes, and compare to execution of
+        the next command. The next command depends on the specifics of the
+        current command, arguments and conditions. We document all commands
+        below.
+
+        OP    | BEHAVIOR
+        IDLE  | increment sram_idx and increment jt_idx
+        NOP   | increment sram_idx and increment jt_idx
+        CHECK | move sram_idx and jt_idx OR increment sram_idx and jt_idx
+        JUMP  | move sram_idx and jt_idx
+        CYCLE | move sram_idx and jt_idx OR increment sram_idx and jt_idx
+        END   | no next entry
+
         :param list[jump_table.JumpEntry] jt_entries: JT entries
         :param list[int] counters: counter values, or None for all 0s
         :param int start_address_ns: SRAM start address, in ns
         :return: jump table object
         :rtype: jump_table.JumpTable
         """
-        for i, a in enumerate(jt_entries):
-            for j, b in enumerate(jt_entries[i+1:]):
-                if abs(a.from_addr - b.from_addr) < cls.JT_MIN_FROM_ADDR_SPACING:
-                    raise ValueError(
-                        "Entries {}({}) and {}({}) have from addrstoo close together.".format(
-                            i, a.operation, i+j+1, b.operation)
-                    )
+
+        # these entries can increment both idx
+        increment_idx = (
+            jump_table.IDLE, jump_table.NOP, jump_table.CHECK,
+            jump_table.CYCLE
+        )
+
+        # these entries can move both idx
+        jump_idx = (
+            jump_table.CHECK, jump_table.JUMP, jump_table.CYCLE
+        )
+
+        # verify no two from_addr are executed within JT_MIN_FROM_ADDR_SPACING
+        for k, jt_entry in enumerate(jt_entries):
+            # first pass we check all increment cases
+            if type(jt_entry.operation) in increment_idx:
+                sram_idx = jt_entry.from_addr
+                next_jt_idx = k + 1
+                next_from_addr = jt_entries[next_jt_idx].from_addr
```

Comment:
Do we check anywhere that the last entry is an END? Or does that get added automatically (probably)? If not it will error here with a confusing message.